subprojects/test-annotation-generator/build.gradle.kts

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+plugins {
+    id("gradlebuild.internal.kotlin")
+}
+
+dependencies {
+    implementation(libs.futureKotlin("reflect"))
+    implementation(libs.futureKotlin("stdlib-jdk8"))
+    implementation(project(":core"))
+    implementation(project(":core-api"))
+    implementation(project(":distributions-full"))
+}
```

Comment:
:question: can't this be a project of `build-logic` instead?

Reply:
Im not sure the build-logic can depend upon `core` and `distribution-full`

Reply:
@jjohannJenks is what @eskatos @is suggesting here possible?

Reply:
I agree that more should go into build-src and there should be no need for this additional subproject. There should be no need to depend on any of the main sources.

Ideally you can just use https://github.com/gradle/gradle/blob/4817230783dcff0d8f2f3f266794c2f454f96959/build-logic/basics/src/main/kotlin/gradlebuild/basics/PublicApi.kt directly to have the default imports. Do you need anything more in addition to the packages?

If more is needed, here is where the file with the imports is generated from that source:
https://github.com/gradle/gradle/blob/4817230783dcff0d8f2f3f266794c2f454f96959/build-logic/packaging/src/main/kotlin/gradlebuild.distributions.gradle.kts#L93-L122

Reply:
Looking at this more closely, I think there are two things we should improve:
- There is no need to have the `AnnotationGenerator` as a separate Java application running in a separate VM. It's also confusing to not have the generator in `build-logic`. This can be solved by moving it there and integrating it directly with the corresponding worker action. (I wonder if we need a worker at all and can not just do it directly in the task action method. This should be a cheap operation.)
- The generation should not always depend on the full distribution. Because it makes the test fixtures, and with that all projects, again depend on the full distro. This will break cacheability and isolation in our build (this is bad!).
At first I thought, we can fix this by generating the annotation for each test set individually instead of a general one. But the annotation needs to be known when we compile the test fixtures.
For now, I would propose to just use the "base" distribution on which the fixtures depend anyway right now. This will already cover a lot of ground. We can improve this later, when we detangle things more in our build. E.g., we could have specific fixtures for specific distributions that live in "distribution-specific" places. Or ideally, if everything that is not base moves out of the distribution, we would have to solve the API and default import thing differently in general. Then we can make sure this thing also works accordingly.

I tried the changes here: https://github.com/gradle/gradle/commit/ececff10cfc715d151cf3f669e33cf2ae0b75f54
You may cherry-pick that to go that way.


An **alternative** solution is to **not** wire the generation into the build, but commit the generated Annotation to git, because it won't change often. Then you can use the full distribution. The generation task can then be made part of "sanityFix" which can run when there is an actual change.